model/DeliveryServerService.php

```diff
@@ -0,0 +1,180 @@
+            $numberOfRepetition = (string) $repeatedDelivery->getOnePropertyValue($numberOfRepetitionProp);
+            $delivery = $repeatedDeliveryService->getParentDelivery($delivery);
+        }
+
+        $deliveryProps = $delivery->getPropertiesValues(array(
+            new core_kernel_classes_Property(TAO_DELIVERY_MAXEXEC_PROP),
+            new core_kernel_classes_Property(TAO_DELIVERY_START_PROP),
+            new core_kernel_classes_Property(TAO_DELIVERY_END_PROP),
+            new core_kernel_classes_Property(DeliveryScheduleService::TAO_DELIVERY_RRULE_PROP),
+        ));
+
+        $propMaxExec = current($deliveryProps[TAO_DELIVERY_MAXEXEC_PROP]);
+        $propStartExec = current($deliveryProps[TAO_DELIVERY_START_PROP]);
+        $propEndExec = current($deliveryProps[TAO_DELIVERY_END_PROP]);
+        $rrule = isset($deliveryProps[DeliveryScheduleService::TAO_DELIVERY_RRULE_PROP]) ? current($deliveryProps[DeliveryScheduleService::TAO_DELIVERY_RRULE_PROP]) : false;
+
+        $settings[TAO_DELIVERY_MAXEXEC_PROP] = (!(is_object($propMaxExec)) or ($propMaxExec=="")) ? 0 : $propMaxExec->literal;
+        $settings[TAO_DELIVERY_START_PROP] = (!(is_object($propStartExec)) or ($propStartExec=="")) ? null : $propStartExec->literal;
+        $settings[TAO_DELIVERY_END_PROP] = (!(is_object($propEndExec)) or ($propEndExec=="")) ? null : $propEndExec->literal;
+        $settings[CLASS_COMPILEDDELIVERY] = $delivery;
+
+        if ($rrule) {
+            $startDate = date_create('@'.$settings[TAO_DELIVERY_START_PROP]);
+            $endDate = date_create('@'.$settings[TAO_DELIVERY_END_PROP]);
+            $diff = date_diff($startDate, $endDate);
+
+            $rule = new \Recurr\Rule((string) $rrule);
+            $transformer = new \Recurr\Transformer\ArrayTransformer();
+            $rEvents = $transformer->transform($rule)->startsBefore(date_create(), true);
+
+            if (isset($repeatedDelivery) && isset($rEvents[$numberOfRepetition])) {
+                $rEvent = $rEvents[$numberOfRepetition];
+                $rEventStartDate = $rEvent->getStart();
+
+                $rEventEndDate = clone $rEvent->getStart();
+                $rEventEndDate->add($diff);
+
+                $settings[TAO_DELIVERY_START_PROP] = $rEventStartDate->getTimestamp();
+                $settings[TAO_DELIVERY_END_PROP] = $rEventEndDate->getTimestamp();
+            } else {
+                foreach ($rEvents as $numberOfRepetition => $rEvent) {
+                    $rEventStartDate = $rEvent->getStart();
+                    $rEventEndDate = clone $rEvent->getStart();
+
+                    $rEventEndDate->add($diff);
+
+                    $repeatedDeliveryExists = ($repeatedDeliveryService->getDelivery($delivery, $numberOfRepetition) !== false);
+                    if ($this->areWeInRange($rEventStartDate, $rEventEndDate) && !$repeatedDeliveryExists) {
```

Comment:
If there is no delivery currently available we display the start and end time of the parent delivery, can we easily get the next available delivery start and end time ?